File: project/jimmer-apt/src/main/java/org/babyfish/jimmer/apt/client/ClientProcessor.java
package org.babyfish.jimmer.apt.client;

import site.addzero.context.Context;
import site.addzero.lsi.apt.diagnostic.AptLsiDiagnostics;
import site.addzero.context.LsiSourceFilterKt;
import site.addzero.lsi.codegen.GeneratedResourceArtifact;
import site.addzero.lsi.jimmer.client.metadata.generator.ClientProcessorSupport;
import site.addzero.lsi.resolver.LsiResolver;

import java.util.Set;

public class ClientProcessor {

    private static final String JIMMER_CLIENT = "META-INF/jimmer/client";

    private final boolean explicitApi;

    public ClientProcessor(
            boolean explicitApi
    ) {
        this.explicitApi = explicitApi;
    }

    public void process() {
        LsiResolver resolver = Context.INSTANCE.getLsiResolver();
        Set<String> serviceTypeNames = ClientProcessorSupport.collectClientSchemaServiceTypeNames(
                resolver,
                Context.INSTANCE.getDelayedClientTypeNames(),
                explicitApi,
                LsiSourceFilterKt::matchesConfiguredSourceFilters
        );
        checkJdkVersion(serviceTypeNames);
        GeneratedResourceArtifact artifact = ClientProcessorSupport.generateClientSchemaArtifact(
                resolver,
                explicitApi,
                serviceTypeNames,
                Context.INSTANCE.guessGeneratedJimmerResourceFile("client"),
                Context.INSTANCE::convertedLsiTypeNameOf,
                Context.INSTANCE::findDraftImplDocMap
        );
        writeArtifact(artifact);
    }

    /**
     * Find this problem on `zulu-1.8 jdk`,
     * `TypeMirror.getAnnotationMirrors` always returns empty list if
     * the current `TypeMirror` is not top type but generic argument.
     */
    private void checkJdkVersion(Set<String> serviceTypeNames) {
        try {
            String.class.getMethod("isBlank");
            return;
        } catch (NoSuchMethodException e) {
            // Do nothing
        }
        // 覆盖来源：project/jimmer-apt/.../client/ClientProcessor.checkJdkVersion
        // 迁移说明：JDK8 `@FetchBy` 防护仍保留在 APT 入口，但服务探测改为复用共享 LSI client API 规则
        if (!serviceTypeNames.isEmpty()) {
            throw new FetchByUnsupportedException();
        }
    }

    private void writeArtifact(GeneratedResourceArtifact artifact) {
        // 覆盖来源：project/jimmer-apt/.../client/ClientProcessor.process 的 resource 输出
        // 迁移说明：APT 入口只负责把 metadata generator 返回的 resource artifact 落盘，
        // resource overwrite 语义统一收口到 LsiFiler adapter，processor 不再直接依赖 APT helper
        try {
            Context.INSTANCE.getLsiFiler().overwriteResourceFile(artifact.getPath(), artifact.getContent());
            Context.INSTANCE.setDelayedClientTypeNames(null);
        } catch (Exception ex) {
            throw AptLsiDiagnostics.generatorException("Cannot write \"" + JIMMER_CLIENT + "\"", ex);
        }
    }

}

File: project/compiler/client/jimmer-ksp-client/src/main/kotlin/org/babyfish/jimmer/ksp/client/ClientProcessor.kt
package org.babyfish.jimmer.ksp.client

import com.google.auto.service.AutoService
import site.addzero.context.Context
import site.addzero.context.Context.delayedClientTypeNames
import site.addzero.context.Settings
import site.addzero.context.matchesConfiguredSourceFilters
import site.addzero.lsi.codegen.GeneratedResourceArtifact
import site.addzero.lsi.jimmer.client.metadata.generator.ClientProcessorSupport
import site.addzero.lsi.jimmer.processor.spi.IMMUTABLE_PROCESSOR
import site.addzero.lsi.jimmer.processor.spi.TYPED_TUPLE_PROCESSOR
import site.addzero.lsi.processor.ProcessorSpi

@AutoService(ProcessorSpi::class)
class ClientProcessor : ProcessorSpi<Context, Unit> {

    override val dependsOn: Set<String>
        get() = setOf(IMMUTABLE_PROCESSOR, TYPED_TUPLE_PROCESSOR)

    private val explicitClientApi get() = Context.explicitClientApi
    private val collectedServiceTypeNames = linkedSetOf<String>()

    override var ctx: Context
        get() = Context
        set(value) {}

    override fun onRound() {
        // 覆盖来源：project/compiler/client/jimmer-ksp-client/.../ClientProcessor.onRound
        // 迁移说明：服务候选收集继续收口到 shared support，processor 壳层只保留 round 生命周期与文件写入
        collectedServiceTypeNames += ClientProcessorSupport.collectClientSchemaServiceTypeNames(
            resolver = ctx.lsiResolver,
            delayedTypeNames = delayedClientTypeNames,
            explicitClientApi = explicitClientApi,
            matchesSourceFilters = { it.matchesConfiguredSourceFilters() }
        )
    }

    override fun onFinish() {
        if (Settings.jimmerBuddyIgnoreResourceGeneration) {
            collectedServiceTypeNames.clear()
            return
        }
        val artifact = ClientProcessorSupport.generateClientSchemaArtifact(
            resolver = ctx.lsiResolver,
            explicitClientApi = explicitClientApi,
            serviceTypeNames = collectedServiceTypeNames,
            existingSchemaFile = ctx.guessGeneratedJimmerResourceFile("client"),
            convertedLsiTypeNameOf = Context::convertedLsiTypeNameOf,
            draftImplDocMapOf = { type, annotationQualifiedName, valueAttributeName ->
                Context.findDraftImplDocMap(type, annotationQualifiedName, valueAttributeName)
            },
        )
        writeArtifact(artifact)
        delayedClientTypeNames = null
        collectedServiceTypeNames.clear()
    }

    private fun writeArtifact(
        artifact: GeneratedResourceArtifact,
    ) {
        // 覆盖来源：project/compiler/client/jimmer-ksp-client/.../ClientProcessor.onFinish resource 输出
        // 迁移说明：resource 写入职责回收到 processor，client metadata-generator 只返回纯 resource artifact
        ctx.lsiFiler.createResourceFile(artifact.path, artifact.content)
    }
}
